pubnub/models/subscription.py

```diff
@@ -0,0 +1,296 @@
+            )
+            self.pubnub._subscription_manager.adapt_subscribe_builder(subscribe_operation)
+
+        return channel_list
+
+    def remove(self, subscription: Subscription) -> list:
+        channel_list = []
+        group_list = []
+
+        if subscription.subscription_type == PNSubscriptionType.CHANNEL:
+            for channel in subscription.subscription_names:
+                if channel in self.channels:
+                    self.channels[channel].remove(subscription)
+                    if self.channels[channel] == []:
+                        del self.channels[channel]
+                        channel_list.append(channel)
+        else:
+            for channel_group in subscription.subscription_names:
+                if channel_group in self.channel_groups:
+                    self.channel_groups[channel_group].remove(subscription)
+                if not self.channel_groups[channel_group]:
+                    del self.channel_groups[channel_group]
+                    group_list.append(channel_group)
+
+        if channel_list:
+            unsubscribe_operation = UnsubscribeOperation(
+                channels=channel_list,
+                channel_groups=group_list
+            )
+            self.pubnub._subscription_manager.adapt_unsubscribe_builder(unsubscribe_operation)
+        return channel_list
+
+    def get_subscribed_channels(self):
+        return list(self.channels.keys())
+
+    def get_subscribed_channel_groups(self):
+        return list(self.channel_groups.keys())
+
+    def get_subscriptions_for(self, subscription_type: PNSubscriptionType, name: str):
+        if subscription_type == PNSubscriptionType.CHANNEL:
+            return [channel for channel in self.get_subscribed_channels() if channel == name]
+        else:
+            return [group for group in self.get_subscribed_channel_groups() if group == name]
+
+    def get_all_listeners(self):
+        listeners = []
+
+        for channel in self.channels:
+            listeners += self.channels[channel]
+        for channel_group in self.channel_groups:
+            listeners += self.channel_groups[channel_group]
+        if self.global_listeners:
+            listeners += self.global_listeners
+        return set(listeners)
+
+    def add_listener(self, listener):
+        assert isinstance(listener, SubscribeCallback)
+        self.global_listeners.append(listener)
+
+    def remove_listener(self, listener):
+        assert isinstance(listener, SubscribeCallback)
+        self.global_listeners.remove(listener)
+
+    def unsubscribe_all(self):
+        unsubscribe_operation = UnsubscribeOperation(
+            channels=self.channels,
+            channel_groups=self.channel_groups
+        )
+        self.pubnub._subscription_manager.adapt_unsubscribe_builder(unsubscribe_operation)
+        self.channels = []
+        self.channel_groups = []
+
+    def unsubscribe(self, channels=None, groups=None):
```

Comment:
there shouldn't be a method like this, you can only subscribe() or unsubscribe() the whole set. If you want to deal with individual subscriptions, use subscription objects

Reply:
it's a facade in `SubscriptionRegistry` that cleans the subscription list and forwards changes to `SubscriptionManager`. I know it's a bit of spaghetti but for 'reasons' I don't want to refactor it right now